### PR TITLE
Support for API tokens

### PR DESCRIPTION
**Pull request by donker**
Implements #5611 

Replaces #5730

### CODE REVIEW

File: DNN Platform/DotNetNuke.Web/Api/Auth/ApiTokens/ApiTokenController.cs [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+            {
+                portalId = Null.NullInteger;
+            }
+
+            string newToken;
+            using (var generator = RandomNumberGenerator.Create())
+            {
+                var tokenBytes = new byte[32];
+                generator.GetBytes(tokenBytes);
+                newToken = Convert.ToBase64String(tokenBytes);
```
Comment (bdukes) [inline DNN Platform/DotNetNuke.Web/Api/Auth/ApiTokens/ApiTokenController.cs:132]: I think we need to use the same conversion here (from `byte[]` to `string`) that we're using during hashing.

```suggestion
                newToken = EncodeBase64(tokenBytes);
```

File: DNN Platform/DotNetNuke.Web/Api/Auth/ApiTokens/ApiTokenSettingsRepository.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information
+
+namespace DotNetNuke.Web.Api.Auth.ApiTokens
+{
+    using DotNetNuke.Entities.Modules.Settings;
+
+    public class ApiTokenSettingsRepository : SettingsRepository<ApiTokenSettings>
```
Comment (bdukes) [inline DNN Platform/DotNetNuke.Web/Api/Auth/ApiTokens/ApiTokenSettingsRepository.cs:9]:
```suggestion
    /// <summary>A repository for retrieving API Token settings.</summary>
    public class ApiTokenSettingsRepository : SettingsRepository<ApiTokenSettings>
```

File: DNN Platform/DotNetNuke.Web/Api/Auth/ApiTokens/ApiTokenTimespan.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+namespace DotNetNuke.Web.Api.Auth.ApiTokens
+{
+    public enum ApiTokenTimespan
+    {
+        Days30,
+        Days60,
+        Days90,
+        Days180,
+        Years1,
+        Years2,
```
Comment (bdukes) [inline DNN Platform/DotNetNuke.Web/Api/Auth/ApiTokens/ApiTokenTimespan.cs:14]:
```suggestion
    /// <summary>The length of time an API token is valid.</summary>
    public enum ApiTokenTimespan
    {
        /// <summary>30 days.</summary>
        Days30,

        /// <summary>60 days.</summary>
        Days60,

        /// <summary>90 days.</summary>
        Days90,

        /// <summary>180 days.</summary>
        Days180,

        /// <summary>One year.</summary>
        Years1,

        /// <summary>Two years.</summary>
        Years2,
```

File: DNN Platform/DotNetNuke.Web/Api/Auth/ApiTokens/Repositories/ApiTokenRepository.cs [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+            }
+        }
+
+        /// <inheritdoc/>
+        public void SetApiTokenLastUsed(ApiTokenBase apiToken)
+        {
+            using (var context = DataContext.Instance())
+            {
+                var sql = @"UPDATE {databaseOwner}{objectQualifier}ApiTokens
+SET LastUsedOnDate=GETDATE()
```
Comment (bdukes) [inline DNN Platform/DotNetNuke.Web/Api/Auth/ApiTokens/Repositories/ApiTokenRepository.cs:178]: The documentation states that this is a UTC date.

```suggestion
SET LastUsedOnDate=GETUTCDATE()
```

File: DNN Platform/Website/Providers/DataProviders/SqlDataProvider/DotNetNuke.Data.SqlDataProvider [hunk 5]
```diff
@@ -210,7 +210,12 @@ DECLARE @CreateOn datetime = {ts '2000-01-01 00:00:00'}
...
+INSERT INTO {databaseOwner}[{objectQualifier}EventLogTypes] ([LogTypeKey], [LogTypeFriendlyName], [LogTypeDescription], [LogTypeOwner], [LogTypeCSSClass]) VALUES (N'APITOKEN_CREATED', N'API Token Created', N'', N'DotNetNuke.Logging.EventLogType', N'GeneralAdminOperation')
+INSERT INTO {databaseOwner}[{objectQualifier}EventLogTypes] ([LogTypeKey], [LogTypeFriendlyName], [LogTypeDescription], [LogTypeOwner], [LogTypeCSSClass]) VALUES (N'APITOKEN_DELETED', N'API Token Deleted', N'', N'DotNetNuke.Logging.EventLogType', N'GeneralAdminOperation')
+INSERT INTO {databaseOwner}[{objectQualifier}EventLogTypes] ([LogTypeKey], [LogTypeFriendlyName], [LogTypeDescription], [LogTypeOwner], [LogTypeCSSClass]) VALUES (N'APITOKEN_REVOKED', N'API Token Revoked', N'', N'DotNetNuke.Logging.EventLogType', N'GeneralAdminOperation')
+INSERT INTO {databaseOwner}[{objectQualifier}EventLogTypes] ([LogTypeKey], [LogTypeFriendlyName], [LogTypeDescription], [LogTypeOwner], [LogTypeCSSClass]) VALUES (N'APPLICATION_SHUTTING_DOWN', N'Application Shutting Down', N'', N'DotNetNuke.Logging.EventLogType', N'GeneralAdminOperation')
```
Comment (bdukes) [inline DNN Platform/Website/Providers/DataProviders/SqlDataProvider/DotNetNuke.Data.SqlDataProvider:218]: Was this one supposed to be added? It doesn't appear to be related to API tokens…
```suggestion
```

**Review by bdukes — APPROVED** [review]

**Review by mitchelsellers — APPROVED** [review]
